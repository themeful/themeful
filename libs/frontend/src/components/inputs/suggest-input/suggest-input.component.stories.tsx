import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import Fragment from 'stencil-fragment'

export default {
  title: 'Components/Inputs/Suggest Input',
  args: {
    samples: [
      {
        label: 'Without items',
      },
      {
        label: 'With items',
        items: ['First Value', 'Second Value'],
      },
      {
        label: 'Min Length',
        value: 'Some Value',
        items: ['First Value', 'Second Value'],
        'min-length': 10,
      },
      {
        label: 'Extra Validation',
        value: '123',
        items: ['First Value', 'Second Value'],
        validation: (value) => (Number(value) > 0 ? null : 'Please enter a number'),
      },
    ],
  },
}

export const suggestInput = ({ samples }) => {
  return (
    <div>
      <div class="header">
        <h1>Suggest Input</h1>
      </div>
      <div
        class="content"
        style={{
          display: 'grid',
          gridTemplateColumns: '300px 300px',
          gridAutoRows: '100px',
        }}
      >
        {samples.map((args) => (
          <Fragment>
            <div class="tf-light" style={{ background: '#fff', padding: '1rem' }}>
              <tf-suggest-input {...args} onInputChange={action('change')} />
            </div>
            <div style={{ padding: '1rem' }}>
              <tf-suggest-input {...args} onInputChange={action('change')} />
            </div>
          </Fragment>
        ))}
      </div>
    </div>
  )
}
