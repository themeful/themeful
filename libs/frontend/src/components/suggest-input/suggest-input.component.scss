:host {
  display: block;
}

.suggest-input {
  display: block;
  color: var(--tf-font-color);

  &__label {
    font: var(--tf-small-font);
    width: 100%;
  }

  &__hint {
    font: var(--tf-small-font);
    display: block;
    min-height: 0.875rem;
    width: 100%;
    color: #ff6464;
    margin: 1px 0 4px;
  }

  &__input {
    font: var(--tf-default-font);
    width: 100%;
    color: var(--tf-font-color);
    border: 0;
    background-color: rgba(128, 128, 128, 0.1);
    border-bottom: 1px solid currentColor;
    box-sizing: border-box;

    &:focus {
      outline: 0 !important;
      border-color: #325af0;
    }
  }

  &--error &__input {
    border-color: #ff6464;
  }

  &--valid &__input {
    border-color: #23c223;
  }
}
