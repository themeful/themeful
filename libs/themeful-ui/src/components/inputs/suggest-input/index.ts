export * from './suggest-input.component'
