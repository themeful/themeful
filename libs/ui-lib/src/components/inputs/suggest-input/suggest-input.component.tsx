import { Component, Event, EventEmitter, h, Method, Prop, State, Watch } from '@stencil/core'

@Component({
  tag: 'tf-suggest-input',
  styleUrl: 'suggest-input.component.scss',
  shadow: true,
})
export class SuggestInputComponent {
  private input!: HTMLInputElement

  /** Input label */
  @Prop() label!: string

  /** Input suggest items */
  @Prop() items: string[] = []

  /** Required input */
  @Prop() required = false

  /** Disabled input */
  @Prop() disabled = false

  /** Min input */
  @Prop() minLength!: number

  /** Max input */
  @Prop() maxLength!: number

  /** Input value */
  @Prop({ mutable: true }) value!: string | number

  /** validation function */
  @Prop() validation!: (input: string | number) => string | null

  /** Input Event */
  @Event({ composed: false }) inputChange!: EventEmitter

  @State() touched = false
  @State() changed = false
  @State() valid!: boolean
  @State() error = ''
  private static counter = 0
  private dataListId!: string

  /** Validate value */
  @Method()
  public validate(): Promise<boolean> {
    this.touched = true
    this.value = this.value || this.input.value
    return Promise.resolve(this.internalValidation())
  }

  /** Changed value */
  @Method()
  public dirty(): Promise<boolean> {
    return Promise.resolve(this.changed)
  }

  @Watch('value')
  public valueChanged(): void {
    if (this.input.value !== this.value) {
      this.input.value = this.value.toString()
    }
  }

  private inputChanged = () => {
    this.changed = true
    this.value = this.input.value
    this.inputChange.emit(this.value)
    if (!this.valid && this.touched) {
      this.internalValidation()
    }
  }

  private blur = (): void => {
    this.internalValidation()
    this.touched = true
  }

  private internalValidation = (): boolean => {
    this.error = (this.validation && this.validation(this.value)) || ''
    if (this.error === '') {
      if (this.required && (this.value === '' || this.value === undefined)) {
        this.error = `This value is required`
      } else if ((this.value?.toString().length || 0) < this.minLength) {
        this.error = `The min length is ${this.minLength}`
      } else if (this.value?.toString().length > this.maxLength) {
        this.error = `The max length is ${this.maxLength}`
      }
    }
    this.valid = this.error === ''
    return this.valid
  }

  public componentWillLoad() {
    SuggestInputComponent.counter += 1
    this.dataListId = `suggestdatalist${SuggestInputComponent.counter}`
  }

  public render(): HTMLTfSuggestInputElement {
    return (
      <label
        class={`suggest-input${this.valid && this.touched ? ' suggest-input--valid' : ''}${
          !this.valid && this.touched ? ' suggest-input--error' : ''
        }`}
      >
        <span class="suggest-input__label">{this.label}</span>
        <input
          ref={(el: HTMLInputElement) => (this.input = el)}
          class="suggest-input__input"
          type="text"
          list={this.dataListId}
          onInput={this.inputChanged}
          onBlur={this.blur}
          autocomplete="off"
          disabled={this.disabled}
          value={this.value}
        />
        <datalist id={this.dataListId}>
          {this.items.map((item) => (
            <option value={item} />
          ))}
        </datalist>
        <p class="suggest-input__hint">{this.error}</p>
      </label>
    )
  }
}
