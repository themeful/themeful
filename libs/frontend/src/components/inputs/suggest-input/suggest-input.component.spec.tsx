import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { SuggestInputComponent } from './suggest-input.component'

describe('SuggestInputComponent', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [SuggestInputComponent],
      autoApplyChanges: true,
      template: () => <tf-suggest-input />,
    })

    expect(page.root).toMatchSnapshot()
  })

  it('renders with values', async () => {
    const items = ['First Value', 'Second Value']
    const page = await newSpecPage({
      components: [SuggestInputComponent],
      autoApplyChanges: true,
      template: () => <tf-suggest-input label="Some Label" value="Test Text" items={items} />,
    })

    expect(page.root).toMatchSnapshot()
  })
})
