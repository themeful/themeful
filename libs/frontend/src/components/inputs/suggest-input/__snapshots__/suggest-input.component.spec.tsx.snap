// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestInputComponent renders 1`] = `
<tf-suggest-input>
  <mock:shadow-root>
    <label class="suggest-input">
      <span class="suggest-input__label"></span>
      <input autocomplete="off" class="suggest-input__input" list="suggestdatalist1" type="text">
      <datalist id="suggestdatalist1"></datalist>
      <p class="suggest-input__hint"></p>
    </label>
  </mock:shadow-root>
</tf-suggest-input>
`;

exports[`SuggestInputComponent renders with values 1`] = `
<tf-suggest-input>
  <mock:shadow-root>
    <label class="suggest-input">
      <span class="suggest-input__label">
        Some Label
      </span>
      <input autocomplete="off" class="suggest-input__input" list="suggestdatalist2" type="text" value="Test Text">
      <datalist id="suggestdatalist2">
        <option value="First Value"></option>
        <option value="Second Value"></option>
      </datalist>
      <p class="suggest-input__hint"></p>
    </label>
  </mock:shadow-root>
</tf-suggest-input>
`;
